fuel-merkle/src/sparse/merkle_tree.rs

```diff
@@ -924,6 +934,56 @@ mod test {
         assert_eq!(hex::encode(root), expected_root);
     }
 
+    #[test]
+    fn test_update_removes_old_entries() {
+        let mut storage = StorageMap::<TestTable>::new();
+        let mut tree = MerkleTree::new(&mut storage);
+        let tenth_index = 9u32;
+
+        for i in 0_u32..tenth_index {
+            let key = key(i.to_be_bytes());
+            tree.update(key, b"DATA").unwrap();
+        }
+        let size_before_tenth = tree.storage().len();
+        let tenth_key = key(tenth_index.to_be_bytes());
+
+        // Given
+        tree.update(tenth_key, b"DATA").unwrap();
+        let size_after_tenth = tree.storage().len();
+        assert_ne!(size_after_tenth, size_before_tenth);
+
+        // When
+        tree.update(tenth_key, b"ANOTHER_DATA").unwrap();
+
+        // Then
+        assert_eq!(tree.storage().len(), size_after_tenth);
```

Comment:
I don't believe that it's guaranteed that the number of entries in the DB will necessarily be the same. Depending on the case, you will end up with more or less placeholder nodes in your path (which will not be in the database) and more or less real nodes. That means that the number of entries is not constant.
Because of the sparseness of the SMT, you're statistically likely to get a path of length 1 or 2 at a low quantity of leaves, and I think that is what is happening here. I will investigate.

Reply:
Discussed with @xgreenx offline re: number of expected nodes during update. We concluded that we can calculate deterministically the number of nodes before and after a consecutive update using the same key. In this way, we can test for a constant value.